### PR TITLE
publish-commit-bottles: use public action

### PR DESCRIPTION
**Pull request by woodruffw**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Adds the 'attestations' permission, which will be required in the future.

~~Also adds a duplicate "last ditch" step that will (hopefully) reduce the number of hard failures we see here, requiring manual rollback of the corresponding package upload.~~

CCing some GH folks to double-check my changes here: @bdehamer @phillmv

### CODE REVIEW

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [inline .github/workflows/publish-commit-bottles.yml:346]: I think this is the right way to continue if this step fails. The other approach is to have the last ditch step use `if: failure()`, which AFAIK means it'll run if _any_ step fails, not just the last one.

**Comment by woodruffw** [comment]
`actionlint` is failing because it doesn't know the `attestations` permission, since it's too new 🙃

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:351]: Something like this might work in case the action is not idempotent:
```suggestion
      - name: Generate build provenance (last ditch)
        if: always() && steps.first-attestation-attempt.conclusion == 'failure'
```
but the first step will need
```
id: first-attestation-attempt
```

**Review by carlocab — APPROVED** [review]

**Comment by carlocab** [comment]
> Edit 2: I see @bdehamer landed this with [rhysd/actionlint@1f0efe1](https://github.com/rhysd/actionlint/commit/1f0efe145326c0886ba32791ffc9d70e12ae6107), it's just not in a release yet.

We could probably apply that patch to the `actionlint` formula (which is what's used in the workflow)

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [reply .github/workflows/publish-commit-bottles.yml:352]: Technically not, since each attestation is a signed timestamp as well. I'll use your suggestion below.

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [reply .github/workflows/publish-commit-bottles.yml:346]: Yeah, agreed. I think the team working on this at GH has been tackling it though internal retries, so I suspect it'll get better over time.

**Comment by woodruffw** [comment]
> We could probably apply that patch to the `actionlint` formula (which is what's used in the workflow)

SGTM! I can do a follow-up PR for that.

**Review by carlocab — APPROVED** [review]
This approach is an improvement, but I feel like the correct thing to do would be to avoid uploading the bottles to GitHub Packages if the build provenance step fails.

This way we can simply re-run publish workflows when needed.

**Comment by woodruffw** [comment]
> This approach is an improvement, but I feel like the correct thing to do would be to avoid uploading the bottles to GitHub Packages if the build provenance step fails.

Yes, agreed. I think the only current blocker for that is splitting the `pr-pull` step into two steps. I'll look into how `brew pr-pull` works and see if I can't adapt things.

File: .github/workflows/publish-commit-bottles.yml
Comment (chenrui333) [reply .github/workflows/publish-commit-bottles.yml:346]: there is no side effect of ignoring the action step failure?

Comment (woodruffw) [reply .github/workflows/publish-commit-bottles.yml:346]: I don't believe so -- if it fails it should fall through to the "last ditch" version, which will then hard-fail the action if it also fails. But maybe I missed something?

**Comment by carlocab** [comment]
> Yes, agreed. I think the only current blocker for that is splitting the `pr-pull` step into two steps. I'll look into how `brew pr-pull` works and see if I can't adapt things.

`brew pr-pull` just calls `brew pr-upload` to do the upload step: https://github.com/Homebrew/brew/blob/020d9947b0a195c9f91b264a0ebf8dc87d6b14d3/Library/Homebrew/dev-cmd/pr-pull.rb#L158-L171

So I think you can just add `--no-upload` to the `pr-pull` call and then add a `pr-upload` step after the build provenance generation step.

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [inline .github/workflows/publish-commit-bottles.yml:299]: I believe these are safe to remove, since this step no longer uploads to GHP.

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [inline .github/workflows/publish-commit-bottles.yml:348]: Is this token required here? I wasn't sure if `pr-upload` needs both the packages token and the normal API token or just the former.

Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:348]: I think you can get rid of it.
```suggestion
```

File: .github/workflows/publish-commit-bottles.yml
Comment (MikeMcQuaid) [inline .github/workflows/publish-commit-bottles.yml:351]:
```suggestion
          set -euo pipefail

          # Don't quote arguments that might be empty; this causes errors.
```

may be necessary to make this fail on failure

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, happy to merge once CI is 🟢 and you're around for a bit to babysit/debug `master` issues. Thanks @woodruffw!

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [reply .github/workflows/publish-commit-bottles.yml:351]: It's actually not a shell thing, it's about the workflow interpolation (`${{ }}`) so that can't be fixed by tweaking the shell.

**Comment by carlocab** [comment]
#171161 should fix the `actionlint` failure.

**Comment by woodruffw** [comment]
> Looks good, happy to merge once CI is 🟢 and you're around for a bit to babysit/debug `master` issues. Thanks @woodruffw!

I'm able to watch this all day today, so whenever is convenient for you and @carlocab as well!

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [inline .github/workflows/publish-commit-bottles.yml:250]: Are the last two still needed? Don't see it in the documentation anymore (https://github.com/github-early-access/generate-build-provenance/commit/205cbc2e63a570c31ee710c84e0b2b906f19b8f5)

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [inline .github/workflows/publish-commit-bottles.yml:347]: Probably also need `HOMEBREW_GPG_PASSPHRASE` here as `pr-upload` generates the bottle commit.

Comment (woodruffw) [reply .github/workflows/publish-commit-bottles.yml:347]: Thanks! In that case I think `pr-pull` doesn't need that, right?

File: .github/workflows/publish-commit-bottles.yml
Comment (woodruffw) [reply .github/workflows/publish-commit-bottles.yml:250]: Yeah, I believe that's right, removing.

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:334]: This (and references to it) needs to move to the `pr-upload` step I think.

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:341]:
```suggestion
        working-directory: ${{steps.pr-pull.outputs.bottle_path}}
```

File: .github/workflows/publish-commit-bottles.yml
Comment (carlocab) [inline .github/workflows/publish-commit-bottles.yml:357]:
```suggestion
          HOMEBREW_GITHUB_PACKAGES_TOKEN: ${{secrets.HOMEBREW_CORE_GITHUB_PACKAGES_TOKEN}}
          REPO_PATH: ${{steps.set-up-homebrew.outputs.repository-path}}
        run: |
          # Don't quote arguments that might be empty; this causes errors when `brew`
          # interprets them as empty arguments when we want `brew` to ignore them instead.
          brew pr-upload \
            --debug \
            --committer="$BREWTESTBOT_NAME_EMAIL" \
            --root-url="https://ghcr.io/v2/homebrew/core" \
            ${{inputs.warn_on_upload_failure && '--warn-on-upload-failure' || ''}} \
            "$PR"

          echo "head_sha=$(git -C "$REPO_PATH" rev-parse HEAD)" >> "$GITHUB_OUTPUT"
```

**Review by carlocab — APPROVED** [review]
I think we should be ready to give this a try after these changes!

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [reply .github/workflows/publish-commit-bottles.yml:343]: Yeah, and likely don't need committer stuff either since we do `--no-cherry-pick --clean`

File: .github/workflows/publish-commit-bottles.yml
Comment (Bo98) [inline .github/workflows/publish-commit-bottles.yml:294]:
```suggestion
```

**Review by carlocab — APPROVED** [review]
Let's try this.

**Comment by ZhongRuoyu** [comment]
It seems that this approach does not work for `all` bottles because the bottle merge step happens after provenance generation.

`HOMEBREW_VERIFY_ATTESTATIONS=1 brew install -d docker-completion`, for example, fails. It can be seen from [the publish job logs](https://github.com/Homebrew/homebrew-core/actions/runs/9116953496/job/25066739947) that attestations were created for the individual bottles before the merge (e.g., `docker-completion--26.1.3.sonoma.bottle.tar.gz`), so there is no attestation for `docker-completion--26.1.3.all.bottle.tar.gz`.

**Comment by woodruffw** [comment]
> It seems that this approach does not work for `all` bottles because the bottle merge step happens after provenance generation.
> 
> 
> 
> `HOMEBREW_VERIFY_ATTESTATIONS=1 brew install -d docker-completion`, for example, fails. It can be seen from [the publish job logs](https://github.com/Homebrew/homebrew-core/actions/runs/9116953496/job/25066739947) that attestations were created for the individual bottles before the merge (e.g., `docker-completion--26.1.3.sonoma.bottle.tar.gz`), so there is no attestation for `docker-completion--26.1.3.all.bottle.tar.gz`.
> 
> 

Huh, I didn't even know about `all` bottles -- this shows how out of date I am 🙂

I'll look into a fix.

**Pull request merged**
